File: poserp/lib/data/services/auth_service.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_endpoints.dart';
import '../models/api_response.dart';
import '../models/role.dart';
import '../models/user.dart';

class AuthService {
  final ApiClient _apiClient;

  AuthService(this._apiClient);

  Future<ApiResponse<User>> login(String email, String password) async {
    final response = await _apiClient.post(
      ApiEndpoints.login,
      data: {'email': email, 'password': password},
    );

    return ApiResponse<User>.fromJson(
      response.data,
      (json) => User.fromJson(json),
    );
  }

  Future<ApiResponse<User>> register({
    required String name,
    required String email,
    required String password,
    String? phone,
    String? role,
  }) async {
    final response = await _apiClient.post(
      ApiEndpoints.register,
      data: {
        'name': name,
        'email': email,
        'password': password,
        if (phone != null && phone.isNotEmpty) 'phone': phone,
        if (role != null && role.isNotEmpty) 'role': role,
      },
    );

    return ApiResponse<User>.fromJson(
      response.data,
      (json) => User.fromJson(json),
    );
  }

  Future<ApiResponse<User>> getMe() async {
    final response = await _apiClient.get(ApiEndpoints.me);

    return ApiResponse<User>.fromJson(
      response.data,
      (json) => User.fromJson(json),
    );
  }

  Future<void> forgotPassword(String email) async {
    await _apiClient.post(
      ApiEndpoints.forgotPassword,
      data: {'email': email},
    );
  }

  Future<ApiResponse<List<User>>> getUsers() async {
    final response = await _apiClient.get(ApiEndpoints.users);
    return ApiResponse<List<User>>.fromJson(
      response.data,
      (json) => (json as List).map((e) => User.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  Future<ApiResponse<User>> createUser(Map<String, dynamic> payload) async {
    final response = await _apiClient.post(
      ApiEndpoints.users,
      data: payload,
    );
    return ApiResponse<User>.fromJson(
      response.data,
      (json) => User.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<ApiResponse<User>> updateUser(String id, Map<String, dynamic> payload) async {
    final response = await _apiClient.put(
      ApiEndpoints.userById(id),
      data: payload,
    );
    return ApiResponse<User>.fromJson(
      response.data,
      (json) => User.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<ApiResponse<dynamic>> deleteUser(String id) async {
    final response = await _apiClient.delete(ApiEndpoints.userById(id));
    return ApiResponse<dynamic>.fromJson(
      response.data,
      (json) => json,
    );
  }

  Future<ApiResponse<List<Role>>> getRoles() async {
    final response = await _apiClient.get(ApiEndpoints.roles);
    return ApiResponse<List<Role>>.fromJson(
      response.data,
      (json) => (json as List).map((e) => Role.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  Future<ApiResponse<Role>> updateRolePermissions(String id, List<String> permissions) async {
    final response = await _apiClient.put(
      ApiEndpoints.roleById(id),
      data: {'permissions': permissions},
    );
    return ApiResponse<Role>.fromJson(
      response.data,
      (json) => Role.fromJson(json as Map<String, dynamic>),
    );
  }
}

File: src/services/authService.ts
import api from "./api";
import type { User, Role } from "@/types";

interface AuthResponse {
  success: boolean;
  data: User;
  token: string;
  message?: string;
}

export const authService = {
  login: async (email: string, password: string): Promise<AuthResponse> => {
    const { data } = await api.post<AuthResponse>("/auth/login", { email, password });
    return data;
  },

  register: async (payload: {
    name: string;
    email: string;
    password: string;
    role?: string;
    phone?: string;
  }): Promise<AuthResponse> => {
    const { data } = await api.post<AuthResponse>("/auth/register", payload);
    return data;
  },

  getMe: async () => {
    const { data } = await api.get("/auth/me");
    return data;
  },

  updateProfile: async (payload: { name: string; phone?: string }) => {
    const { data } = await api.put("/auth/profile", payload);
    return data;
  },

  changePassword: async (payload: { currentPassword?: string; newPassword?: string }) => {
    const { data } = await api.put("/auth/change-password", payload);
    return data;
  },

  getUsers: async (): Promise<{ success: boolean; data: User[] }> => {
    const { data } = await api.get("/auth/users");
    return data;
  },

  updateUser: async (id: string, payload: Partial<User> & { password?: string }): Promise<{ success: boolean; data: User }> => {
    const { data } = await api.put(`/auth/users/${id}`, payload);
    return data;
  },

  createUser: async (payload: Partial<User> & { password?: string }): Promise<{ success: boolean; data: User }> => {
    const { data } = await api.post(`/auth/users`, payload);
    return data;
  },

  deleteUser: async (id: string): Promise<{ success: boolean; message: string }> => {
    const { data } = await api.delete(`/auth/users/${id}`);
    return data;
  },

  getRoles: async (): Promise<{ success: boolean; data: Role[] }> => {
    const { data } = await api.get("/auth/roles");
    return data;
  },

  updateRolePermissions: async (id: string, permissions: string[]): Promise<{ success: boolean; data: Role }> => {
    const { data } = await api.put(`/auth/roles/${id}`, { permissions });
    return data;
  },
};
